### ISSUE TITLE
Rstudio don't show text in panes

### ISSUE THREAD

**Issue by zmalon** [guix-science/guix-science#12]
Hi, first thank you for this channel!
I recently installed Rstudio from this channel, but I can’t use it because the panes don’t show any text. Only in the menu bar the texts appear. See below.

![scrot-2022-01-12_Jan-01-1642017281_1532x821](https://user-images.githubusercontent.com/97608387/149215048-d568f05e-0381-46d7-94cf-f10966fe07a3.png)

I’ve tried to change rendering settings or disable gpu (according Rstudio community for blank screen bug), and install several fonts to default system but the error persists.

I have GuixSD 1.3.

**Comment by PromyLOPh** [comment]
I can reproduce the problem with
```
Generation 99	Jan 10 2022 08:32:00	(current)
  guix 19a4759
    repository URL: https://git.savannah.gnu.org/git/guix.git
    commit: 19a4759cadf601f40d2a32efdbf6190d7341a8ca
```
and `guix shell -L . --pure rstudio r font-dejavu -- rstudio` on a foreign system too. The same problem exists in qutebrowser as well (which also uses qtwebengine), so I’m assuming this is not a bug in this RStudio package. Not sure what to do about that though.

**Comment by hartja1** [comment]
I can confirm the same behavior with Rstudio 2021.09.2+382 (main window appears, but with only a menu bar and no window contents). Could the problem be the resources path? I get many messages in the terminal as follows:

`
2022-02-25T04:50:35.147963Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /gnu/store/ixp8ad9zm6rn7ac6j95s79kq6l3v2qxr-rstudio-2021.09.2+382/bin/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::basic_istream<char> >&) const src/cpp/shared_core/FilePath.cpp:1453; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
2022-02-25T04:50:35.149469Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /gnu/store/ixp8ad9zm6rn7ac6j95s79kq6l3v2qxr-rstudio-2021.09.2+382/bin/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::basic_istream<char> >&) const src/cpp/shared_core/FilePath.cpp:1453; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
`

It seems to be looking in /gnu/store/.../bin/resources/ and the files exist in /gnu/store/.../resources/.

**Comment by PromyLOPh** [comment]
This seems to be a different issue that I cannot reproduce, @hartja1. How do you start RStudio? Do you have any configuration files in well-known locations (i.e. `/etc`)?

**Comment by hartja1** [comment]
Thanks for looking into it, @PromyLOPh.

> How do you start RStudio?

I am starting manually from the command line, in my home directory.
`user@guix ~$ rstudio`

> Do you have any configuration files in well-known locations (i.e. /etc)?

Not that I'm aware of. I took a look in there and didn't see anything suspicious. I'm also on GuixSD, if that helps.

I tried removing ~/.config/rstudio and ~/.config/RStudio just now (both directories existed) and restarted the program, but the same thing happens. I wouldn't imagine those configs were causing the problem, though, because they were default created. I've not been able to successfully launch Rstudio yet.

**Comment by PromyLOPh** [comment]
Could you also `rm -rf ~/.local/share/rstudio`? RStudio saves its session data there and will use any previous installation of RStudio if the session was not terminated. It would also be helpful if you could attach the output of `strace -v -s 4096 -f -o /tmp/rstudio rstudio` written to `/tmp/rstudio`. And finally RStudio Desktop also has a `--run-diagnostics` switch, which may tell us what is wrong.

**Comment by hartja1** [comment]
Ok, I removed the session and config data and collected the strace
and rstudio diagnostics. Please find the results below.

[rstudio.strace.gz](https://github.com/guix-science/guix-science/files/8185402/rstudio.strace.gz)
[diagnostics-report.txt.gz](https://github.com/guix-science/guix-science/files/8185391/diagnostics-report.txt.gz)

The Guix version is:

```
Generation 5	Mar 04 2022 11:24:18	(current)
  guix 1af78ab
    repository URL: https://git.savannah.gnu.org/git/guix.git
    branch: master
    commit: 1af78ab5e80b98d1914b85709c60fa7f9782e1db
  guix-science 8a87e29
    repository URL: https://github.com/guix-science/guix-science.git
    branch: master
    commit: 8a87e29aa06e3378570492bd570bfa39ea5743f7
```

**Comment by PromyLOPh** [comment]
I’m a little puzzled about what is going on. It should never try to load `loading.html` if everything was fine and it definitely does not for me. Unfortunately I don’t have a working machine with GuixSD installed. Can you run rstudio-server correctly?

**Comment by hartja1** [comment]
No, rstudio-server does not seem to be working either. After installing the package and attempting to run, I get the following error:

```
rdebug@guix ~$ rstudio-server start
/home/rdebug/.guix-profile/bin/rstudio-server: line 12: /etc/init.d/rstudio-server: No such file or directory
```

**Comment by PromyLOPh** [comment]
I’m away from my main computer right now, so sorry for the long response time.

We’re starting rstudio like this, which usually works: https://github.com/leibniz-psychology/psychnotebook-app-rstudio/blob/master/startrstudio.in#L49 You don’t need `--www-socket` and `--server-oneshot`, but the dance with config file is necessary I’m afraid.

**Comment by hartja1** [comment]
Thanks for linking the script. I can confirm rstudio-server starts using the script (with the minor wrangling you suggested) and I can view the app in the web browser. I've tested only briefly, but it appears the UI is working properly. There was a minor issue with the font glitching in the Terminal tab, but that was fixed by disabling hardware acceleration in the terminal options. Otherwise, it looks good.

**Comment by wdkrnls** [comment]
I just wanted to comment that I just installed qutebrowser on Guix and have no trouble seeing widgets in it. I am building `rstudio` again right now, which I don't understand since I have tried to use guix-science as a substitute source. The last time I installed `rstudio` I noticed the same blank widget behavior.

**Comment by PromyLOPh** [comment]
Did you authorize the substitute server, @wdkrnls ?

**Comment by wdkrnls** [comment]
@PromyLOPh I thought I followed the instructions. I have the following in my `/etc/config.scm` and a I have reconfigured the system using it: ```	...   (plain-file
	    "guix-science.pub"
	    "(public-key
 (ecc
  (curve Ed25519)
  (q #D4E1CAFAB105581122B326E89804E3546EF905C0D9B39F161BBD8ABB4B11D14A#)
  )
 )
"
	    ))
	  %default-authorized-guix-keys)```

**Comment by wdkrnls** [comment]
Installing rstudio into its own profile without R causes it to error out, complaining about it not being able to find R. Maybe it should be a propagated input? However, as before I still see the same view as shown in the screenshot. There is no visible text outside of the menus. In qutebrowser I do see visible text, both in menus and overlays and textboxes.

**Comment by PromyLOPh** [comment]
Not sure about the substitutes, it works for me. Perhaps it’s the effect of requesting a substitute for the first time:

```console
$ guix time-machine -C channels.scm -- weather rstudio --substitute-urls='https://substitutes.guix.psychnotebook.org'
computing 1 package derivations for x86_64-linux...
looking for 1 store items on https://substitutes.guix.psychnotebook.org...
https://substitutes.guix.psychnotebook.org
  100.0% substitutes available (1 out of 1)
  at least 97.9 MiB of nars (compressed)
  437.8 MiB on disk (uncompressed)
  (continuous integration information unavailable)
```

I don’t know the best way to handle R. On one hand RStudio links to libR.so and needs patches to work with a different R. On the other hand you can (in theory) swap out the R binary and use a different one. Thus propagating R would override this choice.

I can actually get text output if I add `--no-sandbox` to RStudio’s command line flags. I’m guessing the sandbox is blocking font loading somehow…?

edit: Maybe it’s the same issue as https://bugreports.qt.io/browse/QTBUG-92969 ?

**Comment by wdkrnls** [comment]
Great detective work! That QT patch sounds really promising, and even if it doesn't work, `--no-sandbox` sounds like an acceptable compromise to me in order to be able to use rstudio desktop through Guix. :+1:

**Comment by wdkrnls** [comment]
I got excited when I saw the new git commits from a week or two ago which suggested maybe the issue was resolved. My poor little laptop struggled for 2 hours or so to compile rstudio (despite my config accepting binaries?), but unfortunately my initial test revealed that the qt bug remains readily apparent. Of course, I didn't try it in a guix shell. Maybe there it would work? Maybe I just need to reboot and try again. Practical reproducibility still aludes me with Guix. Anyway, I did get a warning with the `rstudio` part (not with QT as far as I could tell). In case it is at all helpful, I have included the text below.

```
     [exec]   → WARNING Statement "index.cjs.js" has failed to resolve in module "w3c-keyname@2.2.4"
     [exec] 
     [exec]   -------------- 
     [exec] 
     [exec] 
     [exec]   -------------- 
     [exec] 
     [exec] Launching quantum core
     [exec]   → Generating abstraction, this may take a while
```
The `quantum core` bit sounds mysterious and a bit sketchy... but I suppose that's RStudio's fault.
